src/main/java/org/broadinstitute/consent/http/service/ElasticSearchService.java

```diff
@@ -0,0 +1,214 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.http.entity.ContentType;
+import org.apache.http.nio.entity.NStringEntity;
+import org.broadinstitute.consent.http.configurations.ElasticSearchConfiguration;
+import org.broadinstitute.consent.http.db.DacDAO;
+import org.broadinstitute.consent.http.db.DataAccessRequestDAO;
+import org.broadinstitute.consent.http.db.DatasetDAO;
+import org.broadinstitute.consent.http.db.UserDAO;
+import org.broadinstitute.consent.http.models.Dataset;
+import org.broadinstitute.consent.http.models.DatasetProperty;
+import org.broadinstitute.consent.http.models.Study;
+import org.broadinstitute.consent.http.models.StudyProperty;
+import org.broadinstitute.consent.http.models.User;
+import org.broadinstitute.consent.http.models.elastic_search.DatasetTerm;
+import org.broadinstitute.consent.http.models.elastic_search.StudyTerm;
+import org.broadinstitute.consent.http.models.elastic_search.UserTerm;
+import org.broadinstitute.consent.http.util.ConsentLogger;
+import org.broadinstitute.consent.http.util.gson.GsonUtil;
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+
+public class ElasticSearchService implements ConsentLogger {
+
+  private final RestClient esClient;
+  private final ElasticSearchConfiguration esConfig;
+  private final DatasetDAO datasetDAO;
+  private final DataAccessRequestDAO dataAccessRequestDAO;
+  private final DacDAO dacDAO;
+  private final UserDAO userDAO;
+  private final UseRestrictionConverter useRestrictionConverter;
+
+  public ElasticSearchService(
+      RestClient esClient,
+      ElasticSearchConfiguration esConfig,
+      UseRestrictionConverter useRestrictionConverter,
+      DatasetDAO datasetDAO,
+      DataAccessRequestDAO dataAccessRequestDAO,
+      DacDAO dacDAO,
+      UserDAO userDAO) {
+    this.esClient = esClient;
+    this.esConfig = esConfig;
+    this.useRestrictionConverter = useRestrictionConverter;
+    this.datasetDAO = datasetDAO;
+    this.dataAccessRequestDAO = dataAccessRequestDAO;
+    this.dacDAO = dacDAO;
+    this.userDAO = userDAO;
+  }
+
+
+  private static final String bulkHeader = """
+      { "index": {"_type": "dataset", "_id": "%d"} }
+      """;
+
+  public Response indexDatasets(List<DatasetTerm> datasets) throws IOException {
+    List<String> bulkApiCall = new ArrayList<>();
+
+    datasets.forEach((dsTerm) -> {
+      bulkApiCall.add(bulkHeader.formatted(dsTerm.getDatasetId()));
+      bulkApiCall.add(GsonUtil.getInstance().toJson(dsTerm) + "\n");
+    });
+
+    Request bulkRequest = new Request(
+        "PUT",
+        "/" + esConfig.getDatasetIndexName() + "/_bulk");
+
+    bulkRequest.setEntity(new NStringEntity(
+        String.join("", bulkApiCall) + "\n",
+        ContentType.DEFAULT_BINARY));
```

Comment:
I think either in the client builder code, we need to add a `Content-Type` header or maybe to specify the content type as json here instead of binary. Caveat, have not tested this out either way.

```
elastic  | [2023-06-01T16:48:35,881][WARN ][o.e.d.r.RestController   ] Content type detection for rest requests is deprecated. Specify the content type using the [Content-Type] header.
```

Reply:
You have to use binary for this endpoint. Don't know why. But when I was testing it, anything else failed.